pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java

```diff
@@ -241,11 +244,179 @@ public boolean isCompatible(ColumnDataType anotherColumnDataType) {
+     * compatible with the type.
+     */
+    public Serializable convert(Object value) {
+      switch (this) {
+        case INT:
+          return ((Number) value).intValue();
+        case LONG:
+          return ((Number) value).longValue();
+        case FLOAT:
+          return ((Number) value).floatValue();
+        case DOUBLE:
+          return ((Number) value).doubleValue();
+        case STRING:
+          return value.toString();
+        case BYTES:
+          return ((ByteArray) value).getBytes();
+        case INT_ARRAY:
+          return (int[]) value;
+        case LONG_ARRAY:
+          if (value instanceof long[]) {
+            return (long[]) value;
+          } else {
+            int[] intValues = (int[]) value;
+            int length = intValues.length;
+            long[] longValues = new long[length];
+            for (int i = 0; i < length; i++) {
+              longValues[i] = intValues[i];
+            }
+            return longValues;
+          }
+        case FLOAT_ARRAY:
+          return (float[]) value;
+        case DOUBLE_ARRAY:
+          if (value instanceof double[]) {
+            return (double[]) value;
+          } else if (value instanceof int[]) {
+            int[] intValues = (int[]) value;
+            int length = intValues.length;
+            double[] doubleValues = new double[length];
+            for (int i = 0; i < length; i++) {
+              doubleValues[i] = intValues[i];
+            }
+            return doubleValues;
+          } else if (value instanceof long[]) {
+            long[] longValues = (long[]) value;
+            int length = longValues.length;
+            double[] doubleValues = new double[length];
+            for (int i = 0; i < length; i++) {
+              doubleValues[i] = longValues[i];
+            }
+            return doubleValues;
+          } else {
+            float[] floatValues = (float[]) value;
+            int length = floatValues.length;
+            double[] doubleValues = new double[length];
+            for (int i = 0; i < length; i++) {
+              doubleValues[i] = floatValues[i];
+            }
+            return doubleValues;
+          }
+        case STRING_ARRAY:
+          return (String[]) value;
+        default:
+          throw new IllegalStateException(String.format("Cannot convert: '%s' to type: %s", value, this));
+      }
+    }
+
+    /**
+     * Formats the value to human-readable format based on the type to be used in the query response.
+     */
+    public Serializable format(Object value) {
+      switch (this) {
+        case BYTES:
+          return BytesUtils.toHexString((byte[]) value);
+        default:
+          return (Serializable) value;
+      }
+    }
+
+    /**
+     * Equivalent to {@link #convert(Object)} and {@link #format(Object)} with a single switch statement.
+     */
+    public Serializable convertAndFormat(Object value) {
+      switch (this) {
```

Comment:
Can we reuse the previous two functions ?

Reply:
For performance reason, I decide to make this extra method with a single switch statement (described in the javadoc). This method will be called on a per-value basis from the query result.